src/control/cmd/daos_agent/procmon.go

```diff
@@ -0,0 +1,212 @@
+	return &procInfo{
+		log:       Log,
+		pid:       Pid,
+		ctx:       child,
+		cancelCtx: cancel,
+		response:  Response,
+	}
+}
+
+func getProcPidInode(pid int32) (uint64, error) {
+
+	pidPath := fmt.Sprintf("/proc/%d", pid)
+	info, err := os.Stat(pidPath)
+	if err != nil {
+		return 0, err
+	}
+
+	// Make sure info's underlying interface is correct.
+	stat, ok := info.Sys().(*syscall.Stat_t)
+	if !ok {
+		return 0, fmt.Errorf("Underlying type of FileInfo.Sys() is not Stat_t")
+	}
+	return stat.Ino, nil
+}
+
+// monitorProcess is the top level interface for monitoring individual processes.
+// the interface is used by procMon to kick off monitoring individual processes
+// under their own child context to allow for terminating individual monitoring routines.
+func (p *procInfo) monitorProcess() {
+
+	Ino, err := getProcPidInode(p.pid)
+	if err != nil {
+		p.response <- NewResponse(p.pid, err)
+		return
+	}
+
+	p.log.Debugf("Monitoring pid:%d\n", p.pid)
+	for {
+		select {
+		case <-p.ctx.Done():
+			return
+		case <-time.After(3 * time.Second):
+			newIno, newErr := getProcPidInode(p.pid)
+			if newErr != nil {
+				if os.IsNotExist(newErr) {
+					p.response <- NewResponse(p.pid, fmt.Errorf("Pid %d terminated unexpectedly", p.pid))
+				} else {
+					p.response <- NewResponse(p.pid, newErr)
+				}
+				return
+			}
+			if newIno != Ino {
+				p.response <- NewResponse(p.pid, fmt.Errorf("Pid %d terminated but another processes took its place %d:%d", p.pid, Ino, newIno))
+				return
+			}
+		}
+	}
+}
+
+// procMon is the top level process monitoring struct which accepts requests to
+// monitor and disconnect processes. Once created it is started by passing a
+// context into the startMonitoring call.
+type procMon struct {
+	log     logging.Logger
+	ctx     context.Context
+	procs   map[int32]*procInfo
+	request chan *procMonRequest
+}
+
+// NewProcMon creates a new process monitor struct setting initializing the
+// internal process map and the request channel.
+func NewProcMon(logger logging.Logger) *procMon {
+	return &procMon{
+		log:     logger,
+		procs:   make(map[int32]*procInfo),
+		request: make(chan *procMonRequest),
+	}
+}
+
+func (p *procMon) handleProcessAttach(ctx context.Context, request *procMonRequest, response chan *procMonResponse) {
+	info, ok := p.procs[request.pid]
+
+	if ok == true {
+		p.log.Errorf("Attempted to monitor process %d more than once", request.pid)
+		return
+	}
+
+	info = NewProcInfo(ctx, p.log, request.pid, response)
+
+	p.procs[request.pid] = info
+	go info.monitorProcess()
+}
+
+func (p *procMon) handleProcessDisconnect(request *procMonRequest) {
+	info, ok := p.procs[request.pid]
+
+	if ok != true {
+		p.log.Errorf("Attempted to disconnect process %d but process is missing", request.pid)
+	}
+
+	info.cancelCtx()
+	delete(p.procs, request.pid)
+	p.log.Debugf("Process %d has disconnected", info.pid)
+}
+
+func (p *procMon) handleRequests(ctx context.Context) {
+
+	response := make(chan *procMonResponse)
```

Comment:
Hm, should this be a member of procMon? Looks like it's only created once. Then it wouldn't need to be passed in to the `p.handleProcess...` calls.

Reply:
This is the return channel from the procInfo monitoring thread and is used to condense all the replies in this handleRequests functions so it ends up being a member of procInfo.

Reply:
Ok that makes sense. I don't anticipate two handleRequests routines running at the same time so this can be put in the top level struct.